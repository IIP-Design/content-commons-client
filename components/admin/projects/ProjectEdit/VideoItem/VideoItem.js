/**
 *
 * VideoItem
 *
 */
import React, { Fragment } from 'react';
import { bool, func, object } from 'prop-types';
import {
  Icon, Loader, Progress
} from 'semantic-ui-react';
import gql from 'graphql-tag';
import { graphql } from 'react-apollo';

import Placeholder from 'components/Placeholder/Placeholder';
import './VideoItem.scss';

/* eslint-disable react/prefer-stateless-function */
class VideoItem extends React.PureComponent {
  constructor( props ) {
    super( props );

    /**
     * @todo simulate upload for dev purposes;
     * replace for production
     * 1MB = 1,048,576 Bytes
     */
    this.MEGABYTE = 1048576;
    this.MIN_INTERVAL = 500;
    this.MAX_INTERVAL = 1500;
    this.MIN_MB_SEC = 400;
    this.MAX_MB_SEC = 500;

    this.state = {
      bytesUploaded: 0,
      nIntervId: null,
      isUploading: false
    };
  }

  componentDidMount = () => {
    /**
     * @todo simulate upload for dev purposes;
     * replace for production
     * min, max interval in milliseconds
     */
    const interval = this.getRandomInt( this.MIN_INTERVAL, this.MAX_INTERVAL );
    const nIntervId = setInterval( this.uploadItem, interval );
    this.setState( { nIntervId } );
  }

  componentWillUnmount = () => {
    clearInterval( this.state.nIntervId );
  }

  /**
   * @todo simulate upload for dev purposes;
   * replace for production
   */
  getRandomInt = ( min, max ) => {
    const minNum = Math.ceil( min );
    const maxNum = Math.floor( max );
    return (
      Math.floor( Math.random() * ( ( maxNum - minNum ) + 1 ) ) + minNum
    );
  }

  getRemainingUnits = ( totalUnits, unitsUploaded ) => (
    totalUnits - unitsUploaded
  )

  getAspectRatio = () => {
    const { thumbnails } = this.props.data.video;
    if ( thumbnails && thumbnails.length ) {
      const { height, width } = thumbnails[0].dimensions;
      return height / width * 100;
    }

    // default to HD aspect ratio
    return 9 / 16 * 100;
  }

  incrementUpload = ( unit, min, max ) => (
    this[unit] * this.getRandomInt( min, max )
  )

  endUpload = intervId => {
    clearInterval( intervId );
  };

  uploadItem = () => {
    /**
     * @todo simulate upload for dev purposes;
     * replace for production
     * min, max increment in megabytes
     */
    this.setState( nextState => {
      const { filesize } = this.props.data.video.files[0];
      const { bytesUploaded } = nextState;

      const remainingBytes = this.getRemainingUnits( filesize, bytesUploaded );
      let increment = this.incrementUpload( 'MEGABYTE', this.MIN_MB_SEC, this.MAX_MB_SEC );

      if ( remainingBytes < increment ) {
        increment = remainingBytes;
      }

      // continue uploading
      if ( bytesUploaded < filesize ) {
        return {
          bytesUploaded: this.state.bytesUploaded + increment,
          isUploading: true
        };
      }

      // stop uploading
      this.endUpload( nextState.nIntervId );
      return { isUploading: false };
    } );
  }

  render() {
    const {
      data: { error, loading, video },
      onClick,
      displayItemInModal
    } = this.props;

    if ( loading ) {
      return (
        <Loader active inline="centered">
          <p style={ { fontSize: '0.75em' } }>
            Preparing file for upload...
          </p>
        </Loader>
      );
    }

    if ( error ) {
      return (
        <li className="item video error" style={ { textAlign: 'center' } }>
          <Icon
            color="red"
            name="exclamation triangle"
            size="large"
          />
          <p style={ { fontSize: '0.75em' } }>
            { `${error ? 'A loading' : 'An uploading'} error occurred for this item.` }
          </p>
        </li>
      );
    }

    if ( !video ) return null;

    const {
      title,
      language,
      thumbnails,
      files
    } = video;

    const hasVideoFiles = files && files.length;
    const hasThumbnails = thumbnails && thumbnails.length;

    const filename = hasVideoFiles ? files[0].filename : 'filename';
    const filesize = hasVideoFiles ? files[0].filesize : 0;

    const alt = hasThumbnails ? thumbnails[0].image.alt : '';
    const url = hasThumbnails ? thumbnails[0].image.url : '';

    const { bytesUploaded, isUploading } = this.state;

    const itemStyle = {
      cursor: isUploading ? 'not-allowed' : 'pointer'
    };
    if ( !displayItemInModal ) itemStyle.cursor = 'default';

    const uploadingClass = isUploading ? ' isUploading' : '';
    const Wrapper = !isUploading && displayItemInModal ? 'button' : 'span';
    const wrapperClass = displayItemInModal ? 'modal-trigger' : 'wrapper';

    return (
      <li className="item video">
        <Wrapper
          className={ wrapperClass }
          onClick={
            !isUploading && displayItemInModal ? onClick : null
          }
          style={ itemStyle }
        >
          <div className={ `thumbnail${uploadingClass}` }>
            <Fragment>
              { hasThumbnails
                ? <img src={ url } alt={ alt } />
                : (
                  <Placeholder
                    parentEl="div"
                    childEl="div"
                    parentStyles={ {
                      position: 'relative',
                      overflow: 'hidden',
                      height: '0',
                      paddingTop: `${this.getAspectRatio()}%`
                    } }
                    childStyles={ {
                      thumbnail: {
                        position: 'absolute',
                        top: '0',
                        left: '0',
                        height: '100%',
                        width: '100%'
                      }
                    } }
                  />
                ) }
              <p className="file-name">{ filename }</p>
            </Fragment>

            { isUploading
              && (
                <div className="loading-animation">
                  <Loader active inline="centered" />
                </div>
              ) }
          </div>

          { isUploading
            && (
              <Progress
                value={ bytesUploaded }
                total={ filesize }
                color="blue"
                size="small"
                active
                progress
                precision={ 0 }
              >
                <p>Upload in progress</p>
              </Progress>
            ) }

          <h3 className={ `item-heading ${language.textDirection}` }>
            { title }
          </h3>
          <p className="item-lang">{ language.displayName }</p>
        </Wrapper>
      </li>
    );
  }
}

VideoItem.propTypes = {
  data: object.isRequired,
  displayItemInModal: bool,
  onClick: func
};

const VIDEO_ITEM_QUERY = gql`
  query VideoItem($id: ID!) {
    video: videoUnit(id: $id) {
      id
      title
      files {
        filename
        filesize
      }
      thumbnails {
        image {
          alt
          url
          dimensions {
            height
            width
          }
        }
      }
      language {
        displayName
        textDirection
      }
    }
  }
`;

export default graphql( VIDEO_ITEM_QUERY, {
  options: props => ( {
    variables: {
      id: props.itemId
    },
  } )
} )( VideoItem );

export { VIDEO_ITEM_QUERY };
