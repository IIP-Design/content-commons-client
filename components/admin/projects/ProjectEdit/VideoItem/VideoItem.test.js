import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import { Icon, Loader } from 'semantic-ui-react';
import VideoItem, { VIDEO_ITEM_QUERY } from './VideoItem';

const props = {
  displayItemInModal: true,
  onClick: jest.fn(),
  itemId: '3728'
};

const mocks = [
  {
    request: {
      query: VIDEO_ITEM_QUERY,
      variables: { id: props.itemId }
    },
    result: {
      data: {
        video: {
          id: 'xyz123',
          title: 'Test Title',
          files: [
            {
              filename: 'test-file.mp4',
              filesize: 662595174
            }
          ],
          thumbnails: {
            image: {
              alt: 'some alt text',
              url: 'https://staticcdp.s3.amazonaws.com/2018/05/courses.america.gov_1481/b3b38d194ff80d06dd837f57a41fe16f.jpg',
              dimensions: {
                height: 393,
                width: 770
              }
            }
          },
          language: {
            displayName: 'English',
            textDirection: 'LTR'
          }
        }
      }
    }
  }
];

const Component = (
  <MockedProvider mocks={ mocks } addTypename={ false }>
    <VideoItem { ...props } />
  </MockedProvider>
);

describe( '<VideoItem />', () => {
  it( 'renders initial loading state without crashing', () => {
    const wrapper = mount( Component );
    const videoItem = wrapper.find( 'VideoItem' );
    const loader = (
      <Loader active inline="centered">
        <p style={ { fontSize: '0.75em' } }>
          Preparing file for upload...
        </p>
      </Loader>
    );

    expect( videoItem.exists() ).toEqual( true );
    expect( videoItem.contains( loader ) ).toEqual( true );
    expect( toJSON( videoItem ) ).toMatchSnapshot();
  } );

  it( 'renders error icon and message if `video` has an error', async () => {
    const errorMocks = [
      {
        request: {
          query: VIDEO_ITEM_QUERY,
          variables: { id: props.itemId }
        },
        result: {
          errors: [{ message: 'There was an error.' }]
        }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={ errorMocks } addTypename={ false }>
        <VideoItem { ...props } />
      </MockedProvider>
    );
    // wait for the data and !loading
    await wait( 0 );
    wrapper.update();

    const videoItem = wrapper.find( 'VideoItem' );
    const li = videoItem.find( 'li.video.error' );
    const icon = <Icon color="red" name="exclamation triangle" size="large" />;

    expect( videoItem.exists() ).toEqual( true );
    expect( li.exists() ).toEqual( true );
    expect( videoItem.contains( icon ) ).toEqual( true );
    expect( videoItem.find( 'p' ).text() )
      .toEqual( 'A loading error occurred for this item.' );
  } );

  it( 'renders `Progress` component if `isUploading`', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const videoItem = () => wrapper.find( 'VideoItem' );
    const progress = () => videoItem().find( 'Progress' );
    const isUploading = () => videoItem().state( 'isUploading' );
    const filesize = () => videoItem().prop( 'data' ).video.files[0].filesize;

    expect( isUploading() ).toEqual( false );
    expect( progress().exists() ).toEqual( isUploading() );

    videoItem().setState( { isUploading: true } );

    expect( isUploading() ).toEqual( true );
    expect( progress().exists() ).toEqual( isUploading() );
    expect( progress().prop( 'total' ) ).toEqual( filesize() );
    expect( toJSON( videoItem() ) ).toMatchSnapshot();
  } );

  it( 'renders null if `video` is null', async () => {
    const nullMocks = [
      {
        request: {
          query: VIDEO_ITEM_QUERY,
          variables: { id: props.itemId }
        },
        result: { data: { video: null } }
      }
    ];

    const wrapper = mount(
      <MockedProvider mocks={ nullMocks } addTypename={ false }>
        <VideoItem { ...props } />
      </MockedProvider>
    );
    await wait( 0 );
    wrapper.update();

    const videoItem = wrapper.find( 'VideoItem' );

    expect( videoItem.html() ).toEqual( null );
  } );

  it( 'renders a `span` component with correct props if `isUploading`', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const videoItem = () => wrapper.find( 'VideoItem' );

    videoItem().setState( { isUploading: true } );

    const span = videoItem().find( 'span' );

    expect( span.exists() ).toEqual( true );
    expect( span.hasClass( 'modal-trigger' ) ).toEqual( true );
    expect( span.prop( 'onClick' ) ).toEqual( null );
    expect( span.prop( 'style' ) )
      .toEqual( { cursor: 'not-allowed' } );
  } );

  it( 'renders a `button` with correct props if `displayItemInModal`', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const videoItem = wrapper.find( 'VideoItem' );
    const button = videoItem.find( 'button' );

    expect( videoItem.prop( 'displayItemInModal' ) ).toEqual( true );
    expect( button.exists() ).toEqual( true );
    expect( button.prop( 'onClick' )._isMockFunction )
      .toEqual( true );
    expect( button.hasClass( 'modal-trigger' ) ).toEqual( true );
    expect( button.prop( 'style' ) )
      .toEqual( { cursor: 'pointer' } );
  } );

  it( 'renders a `span` with correct props if `displayItemInModal` is `false`', async () => {
    const newProps = {
      displayItemInModal: false,
      onClick: jest.fn(),
      itemId: '3728'
    };

    const wrapper = mount(
      <MockedProvider mocks={ mocks } addTypename={ false }>
        <VideoItem { ...newProps } />
      </MockedProvider>
    );
    await wait( 0 );
    wrapper.update();

    const videoItem = () => wrapper.find( 'VideoItem' );
    const span = videoItem().find( 'span' );

    expect( videoItem().prop( 'displayItemInModal' ) ).toEqual( false );
    expect( span.exists() ).toEqual( true );
    expect( span.hasClass( 'wrapper' ) ).toEqual( true );
    expect( span.prop( 'onClick' ) ).toEqual( null );
    expect( span.prop( 'style' ) )
      .toEqual( { cursor: 'default' } );
    expect( toJSON( videoItem() ) ).toMatchSnapshot();
  } );

  it( 'clicking the `button` calls `onClick`', async () => {
    const wrapper = mount( Component );
    await wait( 0 );
    wrapper.update();

    const videoItem = wrapper.find( 'VideoItem' );
    const button = videoItem.find( 'button' );

    button.simulate( 'click' );
    expect( button.exists() ).toEqual( true );
    expect( button.prop( 'onClick' ) ).toHaveBeenCalled();
  } );
} );
