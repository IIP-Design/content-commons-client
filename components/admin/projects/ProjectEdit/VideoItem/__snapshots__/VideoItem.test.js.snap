// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VideoItem /> renders \`Progress\` component if \`isUploading\` 1`] = `
<VideoItem
  data={
    Object {
      "error": undefined,
      "fetchMore": [Function],
      "loading": false,
      "networkStatus": 7,
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "variables": Object {
        "id": "3728",
      },
      "video": Object {
        "files": Array [
          Object {
            "filename": "test-file.mp4",
            "filesize": 662595174,
          },
        ],
        "id": "xyz123",
        "language": Object {
          "displayName": "English",
          "textDirection": "LTR",
        },
        "thumbnails": Object {
          "image": Object {
            "alt": "some alt text",
            "dimensions": Object {
              "height": 393,
              "width": 770,
            },
            "url": "https://staticcdp.s3.amazonaws.com/2018/05/courses.america.gov_1481/b3b38d194ff80d06dd837f57a41fe16f.jpg",
          },
        },
        "title": "Test Title",
      },
    }
  }
  displayItemInModal={true}
  itemId="3728"
  onClick={[MockFunction]}
>
  <li
    className="item video"
  >
    <span
      className="modal-trigger"
      onClick={null}
      style={
        Object {
          "cursor": "not-allowed",
        }
      }
    >
      <div
        className="thumbnail isUploading"
      >
        <Placeholder
          childEl="div"
          childStyles={
            Object {
              "thumbnail": Object {
                "height": "100%",
                "left": "0",
                "position": "absolute",
                "top": "0",
                "width": "100%",
              },
            }
          }
          parentEl="div"
          parentStyles={
            Object {
              "height": "0",
              "overflow": "hidden",
              "paddingTop": "56.25%",
              "position": "relative",
            }
          }
        >
          <div
            className="placeholder"
            style={
              Object {
                "height": "0",
                "overflow": "hidden",
                "paddingTop": "56.25%",
                "position": "relative",
              }
            }
          >
            <div
              key="thumbnail"
              style={
                Object {
                  "backgroundColor": "#d6d7d9",
                  "height": "100%",
                  "left": "0",
                  "marginBottom": "0.625em",
                  "position": "absolute",
                  "top": "0",
                  "width": "100%",
                }
              }
            />
          </div>
        </Placeholder>
        <p
          className="file-name"
        >
          test-file.mp4
        </p>
        <div
          className="loading-animation"
        >
          <Loader
            active={true}
            inline="centered"
          >
            <div
              className="ui active centered inline loader"
            />
          </Loader>
        </div>
      </div>
      <Progress
        active={true}
        color="blue"
        precision={0}
        progress={true}
        size="small"
        total={662595174}
        value={0}
      >
        <div
          className="ui blue small active progress"
          data-percent={0}
        >
          <div
            className="bar"
            style={
              Object {
                "width": "0%",
              }
            }
          >
            <div
              className="progress"
            >
              0%
            </div>
          </div>
          <div
            className="label"
          >
            <p>
              Upload in progress
            </p>
          </div>
        </div>
      </Progress>
      <h3
        className="item-heading LTR"
      >
        Test Title
      </h3>
      <p
        className="item-lang"
      >
        English
      </p>
    </span>
  </li>
</VideoItem>
`;

exports[`<VideoItem /> renders a \`span\` with correct props if \`displayItemInModal\` is \`false\` 1`] = `
<VideoItem
  data={
    Object {
      "error": undefined,
      "fetchMore": [Function],
      "loading": false,
      "networkStatus": 7,
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "variables": Object {
        "id": "3728",
      },
      "video": Object {
        "files": Array [
          Object {
            "filename": "test-file.mp4",
            "filesize": 662595174,
          },
        ],
        "id": "xyz123",
        "language": Object {
          "displayName": "English",
          "textDirection": "LTR",
        },
        "thumbnails": Object {
          "image": Object {
            "alt": "some alt text",
            "dimensions": Object {
              "height": 393,
              "width": 770,
            },
            "url": "https://staticcdp.s3.amazonaws.com/2018/05/courses.america.gov_1481/b3b38d194ff80d06dd837f57a41fe16f.jpg",
          },
        },
        "title": "Test Title",
      },
    }
  }
  displayItemInModal={false}
  itemId="3728"
  onClick={[MockFunction]}
>
  <li
    className="item video"
  >
    <span
      className="wrapper"
      onClick={null}
      style={
        Object {
          "cursor": "default",
        }
      }
    >
      <div
        className="thumbnail"
      >
        <Placeholder
          childEl="div"
          childStyles={
            Object {
              "thumbnail": Object {
                "height": "100%",
                "left": "0",
                "position": "absolute",
                "top": "0",
                "width": "100%",
              },
            }
          }
          parentEl="div"
          parentStyles={
            Object {
              "height": "0",
              "overflow": "hidden",
              "paddingTop": "56.25%",
              "position": "relative",
            }
          }
        >
          <div
            className="placeholder"
            style={
              Object {
                "height": "0",
                "overflow": "hidden",
                "paddingTop": "56.25%",
                "position": "relative",
              }
            }
          >
            <div
              key="thumbnail"
              style={
                Object {
                  "backgroundColor": "#d6d7d9",
                  "height": "100%",
                  "left": "0",
                  "marginBottom": "0.625em",
                  "position": "absolute",
                  "top": "0",
                  "width": "100%",
                }
              }
            />
          </div>
        </Placeholder>
        <p
          className="file-name"
        >
          test-file.mp4
        </p>
      </div>
      <h3
        className="item-heading LTR"
      >
        Test Title
      </h3>
      <p
        className="item-lang"
      >
        English
      </p>
    </span>
  </li>
</VideoItem>
`;

exports[`<VideoItem /> renders initial loading state without crashing 1`] = `
<VideoItem
  data={
    Object {
      "error": undefined,
      "fetchMore": [Function],
      "loading": true,
      "networkStatus": 1,
      "refetch": [Function],
      "startPolling": [Function],
      "stopPolling": [Function],
      "subscribeToMore": [Function],
      "updateQuery": [Function],
      "variables": Object {
        "id": "3728",
      },
    }
  }
  displayItemInModal={true}
  itemId="3728"
  onClick={[MockFunction]}
>
  <Loader
    active={true}
    inline="centered"
  >
    <div
      className="ui active text centered inline loader"
    >
      <p
        style={
          Object {
            "fontSize": "0.75em",
          }
        }
      >
        Preparing file for upload...
      </p>
    </div>
  </Loader>
</VideoItem>
`;
